blender/arm/nodes_logic.py

```diff
@@ -278,101 +280,153 @@ def execute(self, context):
     for tree in bpy.data.node_groups:
         if tree.bl_idname == "ArmLogicTreeType":
             for node in tree.nodes:
-                if node.bl_idname in get_replaced_nodes():
-                    print("Replacing "+ node.bl_idname+ " in Tree "+tree.name)
-                    replace(tree, node)
-        
-    
+                if not isinstance(type(node).arm_version, int):
+                    continue  # TODO: that's a line to remove when all node classes will have their own version set.
+                if not node.is_registered_node_type():
+                    # node type deleted. That's unusual. Or it has been replaced for a looong time.
+                    list_of_errors.add( ('unregistered', None, tree.name) )
+                if node.arm_version < type(node).arm_version:
+                    try:
+                        print(f"Replacing {node.bl_idname} in tree {tree.name}")
+                        replace(tree, node)
+                    except LookupError as err:
+                        list_of_errors.add( ('update failed', node.bl_idname, tree.name) )
+                    except Exception as err:
+                        list_of_errors.add( ('misc.', node.bl_idname, tree.name) )
+                elif node.arm_version > type(node).arm_version:
+                    list_of_errors.add(  ('future version', node.bl_idname, tree.name) )
+
+    # if possible, make a popup about the errors.
+    # also write an error report.
+    if len(list_of_errors) > 0:
+        print('there were errors in node replacement')
+        basedir = os.path.dirname(bpy.data.filepath)
+        reportfile = os.path.join(
+            basedir, 'node_update_failure.{:s}.txt'.format(
+                time.strftime("%Y-%m-%dT%H-%M-%S%z")
+            )
+        )
+        reportf = open(reportfile, 'w')
+        for error_type, node_class, tree_name in list_of_errors:
+            if error_type == 'unregistered':
+                print(f"A node whose class doesn't exist was found in node tree \"{tree_name}\"", file=reportf)
+            elif error_type == 'update failed':
+                print(f"A node of type {node_class} in tree \"{tree_name}\" failed to be updated, "
+                      f"because update isn't implemented (anymore?) for this version of the node", file=reportf)
+            elif error_type == 'future version':
+                print(f"A node of type {node_class} in tree \"{tree_name}\" seemingly comes from a future version of armory. "
+                      f"Please check whether your version of armory is up to date", file=reportf)
+            elif error_type == 'misc.':
+                print(f"", file=reportf)
+            else:
+                print(f"Whoops, we don't know what this error type (\"{error_type}\") means. You might want to report a bug here. "
+                      f"All we know is that it comes form a node of class {node_class} in the node tree called \"{tree_name}\".", file=reportf)
```

Comment:
Very nice!

Reply:
did... did I leave the 'misc.' error message empty?

Reply:
Oh, I didn't noticed that^^ My comment was about the message in L403 which is really good.

Reply:
yeah, I understood this was what you were talking about. thanks!
anyway, I fixed it now.